File: Codility/8_leader/equileader.cpp
// https://codility.com/programmers/task/equi_leader/

#include <algorithm>
#include <memory>

std::unique_ptr<int> findLeader(const vector<int>& A) {
    int size = 0;
    int lastValue = 0;
    for (int a : A)
    {
        if (size == 0 || a == lastValue)
        {
            ++size;
            lastValue = a;
        }
        else
        {
            --size;
        }
    }

    if (size == 0)
        return nullptr;

    return std::count(A.begin(), A.end(), lastValue) > A.size() / 2 ? 
        std::unique_ptr<int>(new int(lastValue)) : 
        nullptr;
    
}

int solution(vector<int> &A) {
    std::unique_ptr<int> leader = findLeader(A);
    if (!leader)
        return 0;

    int leaderCount = 0;
    vector<int> leaderCounts;
    for (int a : A)
    {
        if (a == *leader)
            ++leaderCount;
        leaderCounts.push_back(leaderCount);
    }
    
    int leaderReverseCount = 0;
    int equileaderCount = 0;
    for (size_t i = A.size() - 1; i > 0; --i)
    {
        if (A[i] == *leader)
            ++leaderReverseCount;
            
        if (leaderReverseCount > (A.size() - i) / 2 && leaderCounts[i - 1] > i / 2)
            ++equileaderCount;
    }
    return equileaderCount;
}

File: Codility/8_leader/equileader.py
# https://codility.com/programmers/task/equi_leader/

def find_leader(A):
    size = 0
    for a in A:
        if size == 0 or last_value == a:
            size += 1
            last_value = a
        else:
            size -= 1
    
    if size == 0:
        return None
    
    return last_value if A.count(last_value) > len(A) / 2 else None

def solution(A):
    leader = find_leader(A)
    if leader == None:
        return 0
        
    counts = []
    count = 0
    for a in A:
        if a == leader:
            count += 1
        counts.append(count)
        
    reverse_count = 0
    equileader_count = 0
    for i in xrange(1, len(A)):
        if A[-i] == leader:
            reverse_count += 1
        
        if counts[-i-1] > (len(A) - i) / 2 and reverse_count > i / 2:
            equileader_count += 1
            
    return equileader_count